xds/internal/balancer/ringhash/ringhash.go

```diff
@@ -65,450 +71,257 @@ func (bb) ParseConfig(c json.RawMessage) (serviceconfig.LoadBalancingConfig, err
+		b.shouldRegenerateRing = true
 	}
-
-	// Successful resolution; clear resolver error and return nil.
-	b.resolverErr = nil
+	b.config = newConfig
+	b.mu.Unlock()
 	return nil
 }
 
 func (b *ringhashBalancer) ResolverError(err error) {
-	b.resolverErr = err
-	if b.subConns.Len() == 0 {
-		b.state = connectivity.TransientFailure
-	}
-
-	if b.state != connectivity.TransientFailure {
-		// The picker will not change since the balancer does not currently
-		// report an error.
-		return
-	}
-	b.regeneratePicker()
-	b.cc.UpdateState(balancer.State{
-		ConnectivityState: b.state,
-		Picker:            b.picker,
-	})
+	b.child.ResolverError(err)
 }
 
 func (b *ringhashBalancer) UpdateSubConnState(sc balancer.SubConn, state balancer.SubConnState) {
 	b.logger.Errorf("UpdateSubConnState(%v, %+v) called unexpectedly", sc, state)
 }
 
-// updateSubConnState updates the per-SubConn state stored in the ring, and also
-// the aggregated state.
-//
-//	It triggers an update to cc when:
-//	- the new state is TransientFailure, to update the error message
-//	  - it's possible that this is a noop, but sending an extra update is easier
-//	    than comparing errors
-//
-//	- the aggregated state is changed
-//	  - the same picker will be sent again, but this update may trigger a re-pick
-//	    for some RPCs.
-func (b *ringhashBalancer) updateSubConnState(sc balancer.SubConn, state balancer.SubConnState) {
-	s := state.ConnectivityState
-	if logger.V(2) {
-		b.logger.Infof("Handle SubConn state change: %p, %v", sc, s)
-	}
-	scs, ok := b.scStates[sc]
-	if !ok {
-		b.logger.Infof("Received state change for an unknown SubConn: %p, %v", sc, s)
+func (b *ringhashBalancer) updatePickerLocked() {
+	if b.inhibitChildUpdates {
 		return
 	}
-	oldSCState := scs.effectiveState()
-	scs.setState(s)
-	newSCState := scs.effectiveState()
-	b.logger.Infof("SubConn's effective old state was: %v, new state is %v", oldSCState, newSCState)
-
-	b.state = b.csEvltr.recordTransition(oldSCState, newSCState)
-
-	switch s {
-	case connectivity.TransientFailure:
-		// Save error to be reported via picker.
-		b.connErr = state.ConnectionError
-	case connectivity.Shutdown:
-		// When an address was removed by resolver, b called Shutdown but kept
-		// the sc's state in scStates. Remove state for this sc here.
-		delete(b.scStates, sc)
-	}
-
-	if oldSCState != newSCState {
-		// Because the picker caches the state of the subconns, we always
-		// regenerate and update the picker when the effective SubConn state
-		// changes.
-		b.regeneratePicker()
-		b.logger.Infof("Pushing new state %v and picker %p", b.state, b.picker)
-		b.cc.UpdateState(balancer.State{ConnectivityState: b.state, Picker: b.picker})
-	}
+	state := b.aggregatedStateLocked()
 
-	switch b.state {
-	case connectivity.Connecting, connectivity.TransientFailure:
+	// Start connecting to new endpoints if necessary.
```

Comment:
Earlier this reconnecting logic was *after* we call `UpdateState` on the parent, but it is *before*. Do you think that can cause any subtle behavior changes?

Reply:
I made another change here to only connect an IDLE endpoint when a previous endpoint enters TF. I initially omitted this check because I wasn't sure there were situations when it could return false, while there were IDLE endpoints. I found the scenario when a READY SubConn changes to IDLE. I added a test case for this too.